Content.Shared/Doors/Systems/SharedDoorSystem.cs

```diff
@@ -0,0 +1,583 @@
+            return false;
+
+        door.Partial = true;
+
+        // Make sure no entity waled into the airlock when it started closing.
+        if (!CanClose(uid, door))
+        {
+            door.NextStateChange = GameTiming.CurTime + door.OpenTimeTwo;
+            door.State = DoorState.Opening;
+            door.Dirty();
+            UpdateAppearance(uid, door);
+            return false;
+        }
+
+        physics.CanCollide = true;
+        door.NextStateChange = GameTiming.CurTime + door.CloseTimeTwo;
+        _activeDoors.Add(door);
+        door.Dirty();
+
+        if (door.Occludes && TryComp(uid, out OccluderComponent? occluder))
+            occluder.Enabled = true;
+
+        // Crush any entities. Note that we don't check airlock safety here. This should have been checked before
+        // the door closed.
+        Crush(uid, door, physics);
+        return true;
+    }
+    #endregion
+
+    #region Collisions
+    /// <summary>
+    /// Crushes everyone colliding with us by more than <see cref="IntersectPercentage"/>%.
+    /// </summary>
+    public void Crush(EntityUid uid, DoorComponent? door = null, PhysicsComponent? physics = null)
+    {
+        if (!Resolve(uid, ref door))
+            return;
+
+        // is this door capable of crushing? NOT the same as an airlock safety check. The door will still close.
+        if (!door.CanCrush)
+            return;
+
+        // Crush
+        var stunTime = door.DoorStunTime + door.OpenTimeOne;
+        foreach (var entity in GetColliding(uid, physics))
+        {
+            door.CurrentlyCrushing.Add(entity);
+            if (door.CrushDamage != null)
+                _damageableSystem.TryChangeDamage(entity, door.CrushDamage);
+
+            if (TryComp(entity, out StatusEffectsComponent? status))
+                _stunSystem.TryParalyze(entity, stunTime, true, status);
+        }
+
+        if (door.CurrentlyCrushing.Count == 0)
+            return;
+
+        // queue the door to open so that the player is no longer stunned once it has FINISHED opening.
+        door.NextStateChange = GameTiming.CurTime + door.DoorStunTime;
+        door.Partial = false;
+    }
+
+    /// <summary>
+    ///     Get all entities that collide with this door by more than <see cref="IntersectPercentage"/> percent.
+    public IEnumerable<EntityUid> GetColliding(EntityUid uid, PhysicsComponent? physics = null)
+    {
+        if (!Resolve(uid, ref physics))
+            yield break;
+
+        var doorAABB = physics.GetWorldAABB();
+
+        foreach (var body in _physicsSystem.GetCollidingEntities(Transform(uid).MapID, doorAABB))
+        {
+            // static bodies (e.g., furniture) shouldn't stop airlocks/windoors from closing.
+            if (body.BodyType == BodyType.Static)
+                continue;
+
+            if (body.GetWorldAABB().IntersectPercentage(doorAABB) < IntersectPercentage)
+                continue;
+
+            yield return body.Owner;
+        }
```

Comment:
Can you slap a fat TODO on this for me to fix it.

Reply:
Not sure what the main issue / thing-to-be-done was, so just put in a generic one.